README.md
## 前置：你可能并不需要这个包

如果你仅仅想要控制执行 npm scripts 的 node 版本, 可以在 [package.json config][1] 中添加 `engine-strict: true` 实现。

## m-nover

一个简单的用于验证 `node` 版本 和 包管理器(`npm、yarn`)是否符合期望的命令工具

## Features

- 执行 `npm scripts` 前验证 `node` 版本 和 包管理器(`npm、yarn`)是否符合期望
- 期望的 `node` 版本定义遵循 `semantic version` 规范

## 使用场景

- 如果你同时需要维护多个项目，项目运行要求的 `node` 版本 和 包管理器存在差异时，推荐使用。

## 安装

#### npm

` npm install -D m-nover `

#### yarn

` yarn add -D m-nover `

#### pnpm (pnpm 本身支持该能力，无需使用)

<s>`pnpm install -D m-nover`</s>

## 设置目标 node 版本

m-nover 获取目标 `node` 版本的来源有两个：`package.json` 中的 [engines](https://docs.npmjs.com/cli/v9/configuring-npm/package-json/#engines) 和 脚本参数(`arguments`)。

**其中 脚本参数 的优先级要高于 engines**。  
例如：

```javascript
  "scripts": {
    "prestart": "m-nover -t ^18"
  }
  
 or

  "engines": { "node": "^16" }
```

目标 `node` 版本 会使用 *18* 作为实际的 `node` 版本。如果没有脚本参数，会默认使用 `engines` 的设置作为目标版本

## 使用方式

m-nover 目前只提供了 bin 文件，因此可以在 package.json 中的 `npm scripts` 脚本中使用。

1. 通过 && 组合要执行的命令  

     ```javascript
       "scripts": {
         "start": "m-nover && other command",
         "build": "m-nover && other command"
       }
     ```

     当执行 ```npm run start``` 或者 ```npm run build``` 时，m-nover 会首先校验当前执行命令使用的 node 版本和项目实际需要的版本是否一致。  
     **如果不一致，会提示错误信息并阻止后续命令执行。**  
     **如果一致，则会继续执行后续的命令。**

2. 使用 Pre & Post Scripts

    ```javascript
        "scripts": {
          "prestart": "m-nover",
          "start": "other command",
          "prebuild": "m-nover",
          "build": "other command"
        }
     ```

    执行效果和上面的方式是一致的，比较好的一点是不用改动原有的命令。关于 **`Pre & Post Scripts`** 的使用方式，可以参考：[npm scripts][2]

## 参数说明

- `-t`（不推荐）: &ensp;设置 `node` 目标版本, 后跟项目运行实际需要的 `node` 版本。使用方式为 `-t node-version`, 比如: `-t 18`, `-t ^18`, `-t ~18`
- `target`（推荐）: &ensp;同 `-t`, 使用方式为 `target=node-version`, 比如: `target=18`, `target=^18`, `target=~18`
- `en-pm`: &ensp;设置是否启用包管理器验证，默认是开启的，如果想关闭，可以设置`en-pm=false`
- `enable-pm-verify`: &ensp;同 `en-pm`

[1]: https://docs.npmjs.com/cli/v6/using-npm/config?v=true#engine-strict
[2]: https://docs.npmjs.com/cli/v9/using-npm/scripts


CHANGELOG.md

## 2023-08-18
- v1.0.3 修复包管理器验证bug
- v1.0.2 添加包管理器验证
## 2023-07-24
- 修复 yarn 下载时，chalk 作为 devdependency 未被下载的bug
- 完善 readme 信息
- 添加测试用例


validate_pm.js
const {
  PKG_MANAGER_NPM,
  PKG_MANAGER_YARN,
  PKG_MANAGER_PNPM,
  existLockFile,
} = require("./pmlist");
const chalk = require("chalk");
/**
 * @description 检查当前命令实现的包管理器，目前包含了 npm、yarn、pnpm 三种
 * 注意：如果项目之前是通过 pnpm 管理的，因为pnpm本身支持包管理器的判断，因此可以忽略包管理器的检查功能
 */
const detectedWhichPMInuse = (pkgInfo) => {
  let pkgManager = "";
  if (pkgInfo.packageManager) {
    pkgManager = [PKG_MANAGER_NPM, PKG_MANAGER_YARN, PKG_MANAGER_PNPM].find(
      (pm) => pkgInfo.packageManager.startsWith(pm)
    );
  }
  if (existLockFile("package-lock.json")) {
    pkgManager = PKG_MANAGER_NPM;
  }
  if (existLockFile("yarn.lock")) {
    pkgManager = PKG_MANAGER_YARN;
  }
  if (existLockFile("pnpm-lock.yaml")) {
    pkgManager = PKG_MANAGER_PNPM;
  }
  const current_exec_pm = process.env.npm_execpath;
  if (pkgManager && current_exec_pm.indexOf(pkgManager) < 0) {
    console.log(chalk.red("当前运行的包管理器和项目使用的包管理器不一致"));
    console.log("项目使用的包管理器: ", chalk.yellow(pkgManager));
    process.exit(1);
  }
};

module.exports = detectedWhichPMInuse;


utils/hasCommand.js
const commandList = require("./commandList");

const hasCommand = (command, commandArgs = []) => {
    const matched = commandArgs.filter(c => c.includes(command));
    if(matched.length) {
        if (matched[0] === '-t') {
            const ret = commandArgs.indexOf(command) + 1;
            return commandArgs[ret];
        }
        return matched[0].split('=')[1];
    }
    return false;
}

const validCommandArgument = (command, commandArg) => {
    let valid = false;
    switch(command) {
        case commandList["enable-pm-verify"]:
            if (commandArg === 'true' || commandArg === 'false') {
                valid = true;
            }
            break;
    }
    return valid;
}

module.exports.hasCommand = hasCommand;
module.exports.validCommandArgument = validCommandArgument;

index.js
#! /usr/bin/env node

const semver = require("semver");
const chalk = require("chalk");
const path = require("path");
const detectedWhichPMInuse = require('./validate_pm');
const commandList = require("./utils/commandList");
const { validCommandArgument, hasCommand } = require("./utils/hasCommand");

const args = process.argv.slice(2);
const rootDir = process.cwd();

const pkgInfo = require(path.join(rootDir, "package.json"));

const argumentsMapping = {
  targetNode: null,
  verifyPM: null
};

/**
 * @description 如果 package.json 中指定了 node 版本，则会使用该值作为 target node version
 *
 * 关于 engines 的使用，参考 @link https://docs.npmjs.com/cli/v9/configuring-npm/package-json#engines
 */
const getVersionFromPkgInfo = () => {
  if (pkgInfo.engines) {
    argumentsMapping.targetNode = pkgInfo.engines.node;
  }
};

const parseArguments = () => {
  // get from package.json
  getVersionFromPkgInfo();
  // get from arguments
  const commandArgTarget = hasCommand(commandList['-t'], args) || hasCommand(commandList['target'], args);
  if (commandArgTarget) {
    argumentsMapping.targetNode = commandArgTarget;
  }

  const commandArgPM = hasCommand(commandList['en-pm'], args) || hasCommand(commandList['enable-pm-verify'], args);
  if (commandArgPM) {
    argumentsMapping.verifyPM = commandArgPM;
  }
}

/**
 * @description 获取 target node version 信息，有两个来源：package.json 中的 engines 和 脚本参数(arguments)
 *
 * 其中 脚本参数的优先级要高于 engines，例如：
 *
 * "prestart": m-nover -t ^18
 *
 * "engines": { "node": ^16 }
 *
 * target node version 会使用 18
 *
 * 如果没有脚本参数，会默认使用 engines 的设置作为目标版本
 */
const detectedNodeVersionIfMatch = () => {
  if (!argumentsMapping.targetNode) {
    console.log(chalk.red(`未获取目标 node 版本信息，请参考 README.md 设置`));
    process.exit(1);
  }
  compareCurrentAndTarget();
};

const compareCurrentAndTarget = () => {
  const current = process.versions.node;
  const range = semver.validRange(argumentsMapping.targetNode);
  if (!range) {
    console.log(chalk.red(`目标版本信息设置错误，请参考 README.md 设置`));
    process.exit(1);
  }
  const matched = semver.satisfies(current, range);
  if (!matched) {
    console.log(
      chalk.red("当前 node 版本与目标 node 版本不一致，请切换当前 node 版本")
    );
    console.log("当前 node 版本: ", chalk.yellow(current));
    console.log("目标 node 版本: ", chalk.yellow(range));
    process.exit(1);
  }
  console.log(chalk.green("当前 node 版本与目标 node 版本一致"));
};

parseArguments();


detectedNodeVersionIfMatch();

// 默认开启 包管理器 验证，想要关闭，enable-pm-verify false
if (!argumentsMapping.verifyPM || argumentsMapping.verifyPM === 'true') {
  detectedWhichPMInuse(pkgInfo);
}

process.exit(0);


